spin/07-fermyon-ai.md

```diff
@@ -0,0 +1,197 @@
+
+1. Get a yes/no question from the body of the HTTP request
+2. Use the `Llm.infer` function Spin `Llm` library to generate a response to the question
+
+First, update the request handler to get the question from the request body and return an error if
+the body is empty:
+
+```rs
+fn handle_magic_8_ball(req: Request) -> Result<Response> {
+    let body = req.body().as_deref().unwrap_or_default();
+    let question = std::str::from_utf8(&body)?;
+    if question.is_empty() {
+        return Ok(http::Response::builder()
+        .status(200)
+        .header("Content-Type", "application/json")
+        .body(Some("No question provided".into()))?);
+    }
+    ...
+```
+
+Next, let's update the `answer` function to use the LLM instead of pulling a random response from a
+list. First, get the `llm` module from the Spin Rust SDK:
+```rs
+use spin_sdk::llm;
+```
+
+Now, in `answer` use the `llm::infer` function, specifying which model you have configured Spin to
+use and a prompt. We will use the default `Llama2Chat` model, which we will download later. The
+prompt should tell the system what type of responses it should give along with the user provided
+question. Your `answer` function should look similar to the following:
+
+```rs
+fn answer<'a>(question: &'a str) -> Result<String> {
+    let system_prefix = "System:";
+    let user_prefix = "User:";
+    let intro =  format!("{system_prefix} You are acting as an omniscient Magic 8 Ball, generating short responses to yes or no questions. You should be able to give an answer to everything, even if the reply is maybe or to ask again later. If the question is not a yes or no question, reply that they should only ask yes or no questions. Your tone should be expressive yet polite. Always restrict your answers to 10 words or less. NEVER continue a prompt by generating a User question.\n");
+  
+    let prompt = intro + user_prefix + " " + question;
+    // Set model to default Llama2 or the one configured in runtime-config.toml
+    let model = llm::InferencingModel::Llama2Chat;
+    let response = llm::infer(model, &prompt)?.text;
+    let answer = response[system_prefix.len()..].trim();
+    Ok(answer.to_string())
+}
+```
+
+
+> Note: The `llm::infer_with_options` function can be substituted to also pass in configuration to
+> the LLM such as the maximum tokens that should be used in the request.
+
+Before we run our application, we must first download a LLM. The following steps show how to
+download Llama 2, the model used in Fermyon Cloud. This has the benefit of being a stronger model
+and a consistent local to Cloud experience; however, the model is quite large (6GB). Alternatively,
+you can install a smaller LLM model and configure Spin to use it with a runtime config file. Or, you
+can skip this step and test your app by deploying it to Fermyon Cloud and using the model there. The
+model should be placed in the `.spin/llms` directory:
+
+```sh
+cd .spin
```

Comment:
Could we add the instructions to deploy to Cloud here and move the instructions for running locally at the end?

If running in a devcontainer, code space, or an underpowered computer, this will be extremely slow.

Reply:
Done